pkg/apis/serving/v1beta1/predictor.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2020 kubeflow.org.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	// ExactlyOnePredictorViolatedError is a known error message
+	ExactlyOnePredictorViolatedError = "Exactly one of [Custom, ONNX, Tensorflow, Triton, SKLearn, XGBoost] must be specified in PredictorSpec"
+)
+
+// Predictor is an abstraction over machine learning server frameworks
+// +kubebuilder:object:generate=false
+type Predictor interface {
+	GetContainer(modelName string, containerConcurrency int, config *InferenceServicesConfig) *v1.Container
+	Validate() error
+	Default(config *InferenceServicesConfig)
+	GetStorageUri() *string
+}
+
+// PredictorSpec defines the configuration for a predictor,
+// The following fields follow a "1-of" semantic. Users must specify exactly one spec.
+type PredictorSpec struct {
+	// Spec for SKLearn model server
+	SKLearn *SKLearnSpec `json:"sklearn,omitempty"`
+	// Spec for XGBoost model server
+	XGBoost *XGBoostSpec `json:"xgboost,omitempty"`
+	// Spec for TFServing (https://github.com/tensorflow/serving)
+	Tensorflow *TFServingSpec `json:"tensorflow,omitempty"`
+	// Spec for TorchServe (https://pytorch.org/serve)
+	PyTorch *TorchServeSpec `json:"pytorch,omitempty"`
+	// Spec for Triton Inference Server (https://github.com/NVIDIA/triton-inference-server)
+	Triton *TritonSpec `json:"triton,omitempty"`
+	// Spec for ONNX runtime (https://github.com/microsoft/onnxruntime)
+	ONNX *ONNXRuntimeSpec `json:"onnx,omitempty"`
+	// Passthrough Pod fields or specify a custom container spec
+	*CustomPredictor `json:",inline"`
+	// Extensions available in all components
+	ComponentExtensionSpec `json:",inline"`
+}
+
+// PredictorExtensionSpec defines configuration shared across all predictor frameworks
+type PredictorExtensionSpec struct {
+	// This field points to the location of the trained model which is mounted onto the pod.
+	StorageURI *string `json:"storageUri"`
+	// Runtime version of the predictor docker image
+	// +optional
+	RuntimeVersion string `json:"runtimeVersion,omitempty"`
+	// Container enables overrides for the predictor.
+	// Each framework will have different defaults that are populated in the underlying container spec.
+	// +optional
+	v1.Container `json:",inline"`
+}
+
+// GetPredictor returns the framework for the Predictor
+func (p *PredictorSpec) GetPredictor() (Predictor, error) {
```

Comment:
Same thing re: splitting one-of error handling and GetPredictor() -> Predictor logic.